test_conformance/relationals/test_comparisons_fp.cpp

```diff
@@ -0,0 +1,662 @@
+{
+    int size_s = std::snprintf(nullptr, 0, format.c_str(), args...)
+        + 1; // Extra space for '\0'
+    if (size_s <= 0)
+    {
+        throw std::runtime_error("Error during formatting.");
+    }
+    auto size = static_cast<size_t>(size_s);
+    std::unique_ptr<char[]> buf(new char[size]);
+    std::snprintf(buf.get(), size, format.c_str(), args...);
+    return std::string(buf.get(),
+                       buf.get() + size - 1); // We don't want the '\0' inside
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T, typename F> bool verify(const T& A, const T& B)
+{
+    return F()(A, B);
+}
+
+//--------------------------------------------------------------------------
+
+RelationalsFPTest::RelationalsFPTest(cl_device_id device, cl_context context,
+                                     cl_command_queue queue, const char* fn,
+                                     const char* op)
+    : device(device), context(context), queue(queue), fnName(fn), opName(op)
+{
+    // hardcoded for now, to be changed into typeid().name solution in future
+    // for now C++ spec doesn't guarantee human readable type name
+
+    eqTypeNames = { { kHalf, "short" },
+                    { kFloat, "int" },
+                    { kDouble, "long" } };
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T>
+void RelationalsFPTest::generate_equiv_test_data(T* outData,
+                                                 const unsigned int& vecSize,
+                                                 const bool& alpha,
+                                                 const RelTestParams<T>& param,
+                                                 const MTdata& d)
+{
+    unsigned int i;
+
+    generate_random_data(param.dataType, vecSize * TEST_SIZE, d, outData);
+
+    // Fill the first few vectors with NAN in each vector element (or the second
+    // set if we're alpha, so we can test either case)
+    if (alpha) outData += vecSize * vecSize;
+    for (i = 0; i < vecSize; i++)
+    {
+        outData[0] = param.nan;
+        outData += vecSize + 1;
+    }
+    // Make sure the third set is filled regardless, to test the case where both
+    // have NANs
+    if (!alpha) outData += vecSize * vecSize;
+    for (i = 0; i < vecSize; i++)
+    {
+        outData[0] = param.nan;
+        outData += vecSize + 1;
+    }
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T, typename U>
+void RelationalsFPTest::verify_equiv_values(const unsigned int& vecSize,
+                                            const T* const inDataA,
+                                            const T* const inDataB,
+                                            U* const outData,
+                                            const VerifyFunc<T>& verifyFn)
+{
+    unsigned int i;
+    int trueResult;
+    bool result;
```

Comment:
I believe modernized approach is healthier for performance and memory consumption. Are there any reasons to stick to older convention ? It was forced by C compiler once but now CTS project expects C++ compiler anyway, right ?

Reply:
> Are there any reasons to stick to older convention ?

Not really as far as I am aware; we just never had the time to update this throughout the code base after switching to C++.

I'd personally try to avoid adding too many refactoring changes into a PR though. Ewan's suggestion is definitely valuable, but since the code was moved (rather than being entirely new code) I'd encourage doing such improvements in a separate PR. I'm sure we can come up with many more valuable modernization/refactoring changes, but doing too many things in one PR isn't ideal for various reasons (harder to review/easier to miss mistakes, harder to bisect bugs into large commits).

Reply:
Happy to resolve this comment and do the clean-up later